### DISCUSSION TITLE
How do I load a table created by dbt using an IO manager?

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#14648] [Q&A] [answered]
I have an asset that's a table in a database, which is populated using a dbt model.

I have an asset downstream of that asset, which is populated by reading that table into a pandas `DataFrame`, doing some transformations in pandas, and then writing it out elsewhere.

I want to use I/O managers to handle my input-loading: I don't want the body of my `@asset`-decorated function to include code that loads the table. Rather, I want the body of my asset-decorated function to be a pure function that takes a `DataFrame` parameter and returns a `DataFrame` value:

```python
# includes an dbt model/ asset named "raw_table"
dbt_assets = load_assets_from_dbt_manifest(...)

@asset(...)
def transformed_table(raw_table: DataFrame) -> DataFrame:
    ...

defs = Definitions(assets=[*dbt_assets, transformed_table])
```

How do I tell Dagster to load my input `DataFrame` from the database where dbt is storing it?

**Comment by sryza** [comment, marked as answer]
There are two ways to accomplish this:
- Set `io_manager_key` where you define the dbt assets
- Set `input_manager_key` on the asset that depends on the dbt assets

### Set `io_manager_key` where you define the dbt assets

If you have control over the code that loads dbt assets asset definitions, you can set an `io_manager_key` on the dbt assets. The `io_manager_key` won't affect how dbt writes the assets, but it will determine how they're loaded as inputs to assets that depend on them.

For example, if your dbt assets are stored in Snowflake, you might do:

```python
from dagster_snowflake_pandas import SnowflakePandasIOManager

dbt_assets = load_assets_from_dbt_manifest(io_manager_key="snowflake_io_manager", ...)

...

defs = Definitions(
    assets=...,
    resources={"snowflake_io_manager": SnowflakePandasIOManager(database=...)},
)
```

Then, Dagster will use the `SnowflakePandasIOManager` when loading the dbt input to `transformed_table`, without needing to set anything directly on `transformed_table`

The advantage of this approach is that you only need to specify the I/O manager once, rather than on each non-dbt asset that's downstream of your dbt model. IO managers typically correspond to storage systems, so by assigning `io_manager_key="snowflake_io_manager"`, you are essentially telling Dagster "these dbt models correspond to tables in Snowflake".

### Set `input_manager_key` on the asset that depends on the dbt assets

If you don't have control over the code that defines your dbt assets, you can control what I/O manager is used at the site where the assets are loaded:

```python
from dagster_snowflake_pandas import SnowflakePandasIOManager

@asset(ins={"raw_table": AssetIn(input_manager_key="snowflake_io_manager")}, ...)
def transformed_table(raw_table: DataFrame) -> DataFrame:
    ...

defs = Definitions(
    assets=...,
    resources={"snowflake_io_manager": SnowflakePandasIOManager(database=...)},
)
```

The advantage of this approach is that it doesn't require modifying the dbt asset definitions.

**Comment by Kamal-Klm** [comment]
@sryza 
I feel that I may be missing something. I think there is no default IO Manager to read dbt models into pandas dataframes ? Please correct me.
In the docs I came across the input_manager_key "pandas_df_manager". Perhaps its for demonstration only ?

**Reply by sryza** [reply]
Hi @Kamal-Klm - IO managers generally correspond to storage systems.  So e.g. if your dbt models are being saved to Snowflake, then you'll want to use a snowflake IO manager.  Or if they're being saved to DuckDB, you'll want to use a DuckDB IO manager.  The built-in IO managers for these can read tables as Pandas DataFrames.

**Reply by pascal-schetelat** [reply]
I think I have the same question. I want to be able to query a dbt model when I choose to do so : 

```python
@asset(compute_kind='python',
             group_name="API",             
             ins={  "my_dbt_asset": AssetIn(input_manager_key="dbt_to_pandas_df_manager",
                key = get_asset_key_for_model([all_dbt_assets], "my_dbt_asset"),
            )})
def my_aggregate_asset(context):
    return my_dbt_asset.groupby(['A']).sum()
    
```
The IoManager is only a part of the answer; I was looking for such a manager (dagster docs teases a "pandas_df_manager"), to no avail.

I am working on such a manager. It takes as dependancy a `CredentialsResource` 
The hard part to me is  extracting the compiled sql of a dbt model, it feels very hacky to extract it form logs.
